File: common/src/main/java/com/example/common/dto/fund/ApiResponse.java
package com.example.common.dto.fund;

import lombok.*;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ApiResponse<T> {
    private boolean success;
    private T data;
    private String message;
    private String errorCode;
    private PaginationInfo pagination;

    // 성공 응답 생성 메서드들
    public static <T> ApiResponse<T> success(T data) {
        return ApiResponse.<T>builder()
                .success(true)
                .data(data)
                .build();
    }
    public static <T> ApiResponse<T> success(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .data(data)
                .message(message)
                .build();
    }
    public static <T> ApiResponse<T> success(T data, PaginationInfo pagination) {
        return ApiResponse.<T>builder()
                .success(true)
                .data(data)
                .pagination(pagination)
                .build();
    }
    public static <T> ApiResponse<T> success(T data, String message, PaginationInfo pagination) {
        return ApiResponse.<T>builder()
                .success(true)
                .data(data)
                .message(message)
                .pagination(pagination)
                .build();
    }

    // 실패 응답 생성 메서드들
    public static <T> ApiResponse<T> failure(String message) {
        return ApiResponse.<T>builder()
                .success(false)
                .message(message)
                .build();
    }
    public static <T> ApiResponse<T> failure(String message, String errorCode) {
        return ApiResponse.<T>builder()
                .success(false)
                .message(message)
                .errorCode(errorCode)
                .build();
    }
}

File: mobile-front/lib/models/api_response.dart
class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? message;
  final String? errorCode;
  final PaginationInfo? pagination;

  ApiResponse({required this.success, this.data, this.message, this.errorCode, this.pagination});

  factory ApiResponse.fromJson(Map<String, dynamic> j, T Function(Object? json) parseData) {
    return ApiResponse<T>(
      success: j['success'] == true,
      data: j.containsKey('data') ? parseData(j['data']) : null,
      message: j['message'] as String?,
      errorCode: j['errorCode'] as String?,
      pagination: j['pagination'] == null
          ? null
          : PaginationInfo.fromJson(j['pagination'] as Map<String, dynamic>),
    );
  }
}

class PaginationInfo {
  final int page; // 0-based
  final int limit;
  final int totalPages;
  final int currentItems;
  final int total;
  final bool hasNext;
  final bool hasPrev;

  PaginationInfo({
    required this.page,
    required this.limit,
    required this.totalPages,
    required this.currentItems,
    required this.total,
    required this.hasNext,
    required this.hasPrev,
  });

  factory PaginationInfo.fromJson(Map<String, dynamic> j) => PaginationInfo(
    page: j['page'],
    limit: j['limit'],
    totalPages: j['totalPages'],
    currentItems: j['currentItems'],
    total: j['total'],
    hasNext: j['hasNext'],
    hasPrev: j['hasPrev'],
  );
}